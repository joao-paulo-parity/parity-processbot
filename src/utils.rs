use crate::{cmd::*, error::*, github::*, github_bot::GithubBot, Result};

pub struct RepositoryUpdateOutput {
	pub base_sha: String,
	pub head_sha: String,
}

pub enum RepositoryUpdateStrategy {
	FromSubstrateToPolkadotCompanion,
}

pub async fn update_repository(
	github_bot: &GithubBot,
	owner: &str,
	owner_repo: &str,
	contributor: &str,
	contributor_repo: &str,
	contributor_branch: &str,
	update_strategy: Option<RepositoryUpdateStrategy>,
) -> Result<RepositoryUpdateOutput> {
	let token = github_bot.client.auth_key().await?;
	let secrets_to_hide = [token.as_str()];
	let secrets_to_hide = Some(&secrets_to_hide[..]);

	let repo_dir = format!("./{}", owner_repo);
	let (owner_remote_address, owner_repository_domain) =
		github_bot.get_fetch_components(owner, owner_repo, &token);

	if Path::new(&repo_dir).exists() {
		log::info!("{} is already cloned; skipping", &owner_repository_domain);
	} else {
		run_cmd_in_cwd(
			"git",
			&["clone", "-v", &owner_remote_address],
			CommandMessage::Configured(CommandMessageConfiguration {
				secrets_to_hide,
				are_errors_silenced: false,
			}),
		)
		.await?;
	}

	let (contributor_remote_address, _) =
		github_bot.get_fetch_components(contributor, contributor_repo, &token);
	let contributor_remote_branch =
		format!("{}/{}", contributor, contributor_branch);

	// The contributor's remote entry might exist from a previous run (not expected for a fresh
	// clone). If so, delete it so that it can be recreated.
	if run_cmd(
		"git",
		&["remote", "get-url", contributor],
		&repo_dir,
		CommandMessage::Configured(CommandMessageConfiguration {
			secrets_to_hide,
			are_errors_silenced: true,
		}),
	)
	.await
	.is_ok()
	{
		run_cmd(
			"git",
			&["remote", "remove", contributor],
			&repo_dir,
			CommandMessage::Configured(CommandMessageConfiguration {
				secrets_to_hide,
				are_errors_silenced: false,
			}),
		)
		.await?;
	}
	run_cmd(
		"git",
		&["remote", "add", contributor, &contributor_remote_address],
		&repo_dir,
		CommandMessage::Configured(CommandMessageConfiguration {
			secrets_to_hide,
			are_errors_silenced: false,
		}),
	)
	.await?;

	run_cmd(
		"git",
		&["fetch", contributor, contributor_branch],
		&repo_dir,
		CommandMessage::Configured(CommandMessageConfiguration {
			secrets_to_hide,
			are_errors_silenced: false,
		}),
	)
	.await?;

	// The contributor's branch might exist from a previous run (not expected for a fresh clone).
	// If so, delete it so that it can be recreated.
	let _ = run_cmd(
		"git",
		&["branch", "-D", contributor_branch],
		&repo_dir,
		CommandMessage::Configured(CommandMessageConfiguration {
			secrets_to_hide,
			are_errors_silenced: true,
		}),
	)
	.await;
	run_cmd(
		"git",
		&["checkout", "--track", &contributor_remote_branch],
		&repo_dir,
		CommandMessage::Configured(CommandMessageConfiguration {
			secrets_to_hide,
			are_errors_silenced: false,
		}),
	)
	.await?;

	let owner_remote = "origin";
	let owner_branch = "master";
	let owner_remote_branch = format!("{}/{}", owner_remote, owner_branch);

	// Ensure the owner branch has the latest remote changes so that the merge commit is always
	// done with the most updated code from Github.
	run_cmd(
		"git",
		&["fetch", owner_remote, &owner_branch],
		&repo_dir,
		CommandMessage::Configured(CommandMessageConfiguration {
			secrets_to_hide,
			are_errors_silenced: false,
		}),
	)
	.await?;

	// Keep track of the HEAD for the owner's branch *after* fetching the latest changes; in case
	// this reference does not change between attempts, then the master merge commit will have the
	// same effect as the previous attempt, resulting in failure again, thus it's not meaningful to
	// keep retrying in that case.
	let base_sha = {
		let output = run_cmd_with_output(
			"git",
			&["rev-parse", &owner_remote_branch],
			&repo_dir,
			CommandMessage::Configured(CommandMessageConfiguration {
				secrets_to_hide,
				are_errors_silenced: false,
			}),
		)
		.await?;
		String::from_utf8(output.stdout)
			.context(Utf8)?
			.trim()
			.to_string()
	};

	// Create master merge commit before updating packages
	let master_merge_result = run_cmd(
		"git",
		&["merge", &owner_remote_branch, "--no-ff", "--no-edit"],
		&repo_dir,
		CommandMessage::Configured(CommandMessageConfiguration {
			secrets_to_hide,
			are_errors_silenced: false,
		}),
	)
	.await;
	if let Err(e) = master_merge_result {
		log::info!("Aborting repository update due to master merge failure");
		let _ = run_cmd(
			"git",
			&["merge", "--abort"],
			&repo_dir,
			CommandMessage::Configured(CommandMessageConfiguration {
				secrets_to_hide,
				are_errors_silenced: false,
			}),
		)
		.await;
		return Err(e);
	}

	if let Some(update_strategy) = update_strategy {
		match update_strategy {
			RepositoryUpdateStrategy::FromSubstrateToPolkadotCompanion => {
				// `cargo update` should normally make changes to the lockfile with the latest SHAs
				// from Github
				run_cmd(
					"cargo",
					&["update", "-vp", "sp-io"],
					&repo_dir,
					CommandMessage::Configured(CommandMessageConfiguration {
						secrets_to_hide,
						are_errors_silenced: false,
					}),
				)
				.await?;

				// Check if `cargo update` resulted in any changes. If the master merge commit
				// already had the latest lockfile then no changes might have been made.
				let changes_after_update_output = run_cmd_with_output(
					"git",
					&["status", "--short"],
					&repo_dir,
					CommandMessage::Configured(CommandMessageConfiguration {
						secrets_to_hide,
						are_errors_silenced: false,
					}),
				)
				.await?;
				if !String::from_utf8_lossy(
					&(&changes_after_update_output).stdout[..],
				)
				.trim()
				.is_empty()
				{
					run_cmd(
						"git",
						&["commit", "-am", "update Substrate"],
						&repo_dir,
						CommandMessage::Configured(
							CommandMessageConfiguration {
								secrets_to_hide,
								are_errors_silenced: false,
							},
						),
					)
					.await?;
				}
			}
		}
	}

	run_cmd(
		"git",
		&["push", contributor, contributor_branch],
		&repo_dir,
		CommandMessage::Configured(CommandMessageConfiguration {
			secrets_to_hide,
			are_errors_silenced: false,
		}),
	)
	.await?;

	log::info!(
		"Getting the head SHA after a repository update in {}",
		&contributor_remote_branch
	);
	let head_sha = {
		let output = run_cmd_with_output(
			"git",
			&["rev-parse", "HEAD"],
			&repo_dir,
			CommandMessage::Configured(CommandMessageConfiguration {
				secrets_to_hide,
				are_errors_silenced: false,
			}),
		)
		.await?;
		String::from_utf8(output.stdout)
			.context(Utf8)?
			.trim()
			.to_string()
	};

	Ok(RepositoryUpdateOutput { head_sha, base_sha })
}
