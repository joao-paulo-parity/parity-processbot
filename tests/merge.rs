use insta::assert_snapshot;
use parity_processbot::{github, webhook::handle_payload};
use std::fs;

mod utils;

use utils::{cmd::*, common::*, *};

#[tokio::test]
async fn recovers_from_outdated_master() {
	let common_setup_result = &common_setup().await;
	let CommonSetupOutput {
		log_dir,
		git_fetch_url,
		state,
		org,
		user,
		bot_username,
		..
	} = common_setup_result;

	let owner = &create_owner(&common_setup_result);
	let pr_branch = "contributor_patches";

	// Create PR branch
	run_cmd(
		"git",
		&["checkout", "-b", pr_branch],
		&owner.repo_dir,
		Some(CmdConfiguration::SilentStderrStartingWith(&[
			"Switched to a new branch",
		])),
	);
	let pr_head_sha =
		get_cmd_output("git", &["rev-parse", "HEAD"], &owner.repo_dir);

	// Commit on master to make the PR branch outdated
	run_cmd(
		"git",
		&["checkout", owner.branch],
		&owner.repo_dir,
		Some(CmdConfiguration::SilentStderrStartingWith(&[
			"Switched to branch",
		])),
	);
	fs::write(
		owner.repo_dir.join("src").join("main.rs"),
		"fn main() {println!(\"CHANGED!\");}",
	)
	.unwrap();
	run_cmd("git", &["add", "."], &owner.repo_dir, None);
	run_cmd(
		"git",
		&["commit", "-m", "make contributor branch outdated"],
		&owner.repo_dir,
		None,
	);

	let SetupGithubAPIOutput {
		github_api,
		api_root,
	} = &setup_github_api(bot_username, org, vec![user.clone()]);
	setup_coredevs(github_api, org, vec![user.clone()]);
	setup_substrateteamleads(github_api, org, vec![user.clone()]);
	setup_commit(github_api, org, &owner, &pr_head_sha);
	let pr = &setup_pull_request(
		github_api,
		api_root,
		org,
		&owner,
		&pr_head_sha,
		pr_branch,
		2,
	);

	handle_payload(
		github::Payload::IssueComment {
			action: github::IssueCommentAction::Created,
			comment: github::Comment {
				body: "bot merge".to_string(),
				user: Some(owner.user.clone()),
			},
			issue: github::Issue {
				number: pr.number,
				body: None,
				html_url: pr.html_url.clone(),
				repository_url: Some(pr.repository_url.to_string()),
				pull_request: Some(github::IssuePullRequest {}),
				repository: Some(owner.repo.clone()),
				user: Some(owner.user.clone()),
			},
		},
		state,
	)
	.await
	.unwrap();

	assert_snapshot!(read_snapshot(
		(&log_dir).path().to_path_buf(),
		&[git_fetch_url]
	));
}
