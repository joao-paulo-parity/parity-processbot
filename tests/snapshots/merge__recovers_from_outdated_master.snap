---
source: tests/merge.rs
expression: "read_snapshot((&log_dir).path().to_path_buf(), &[git_fetch_url])"

---
INFO [parity_processbot::webhook] Received merge request for PR https://github.com/org/owner/pull/1 from user foo
INFO [parity_processbot::webhook] https://github.com/org/owner/pull/1 is mergeable
INFO [parity_processbot::webhook] https://github.com/org/owner/pull/1 merge requested by a team lead.
INFO [parity_processbot::webhook] CombinedStatus { state: Success, statuses: [Status { id: 1, context: "DOES_NOT_MATTER", state: Success }] }
INFO [parity_processbot::webhook] https://github.com/org/owner/pull/1 has success status
INFO [parity_processbot::webhook] CheckRuns { check_runs: [CheckRun { id: 1, name: "DOES_NOT_MATTER", status: Completed, conclusion: Some(Success), head_sha: "MDEwOlJlcG9zaXRvcnkxMDk4NzI2MjA" }] }
INFO [parity_processbot::webhook] https://github.com/org/owner/pull/1 has successful checks
INFO [parity_processbot::webhook] https://github.com/org/owner/pull/1 checks successful; trying merge.
INFO [parity_processbot::webhook] Attempting to recover from possibly outdated target branch (from org/owner/contributor_patches into org/owner/master)
INFO [parity_processbot::cmd] Run git ["clone", "-v", "{REDACTED}/org/owner"] in {REDACTED}
INFO [parity_processbot::cmd] Run git ["remote", "get-url", "org"] in {REDACTED}
INFO [parity_processbot::cmd] Run git ["remote", "add", "org", "{REDACTED}/org/owner"] in {REDACTED}
INFO [parity_processbot::cmd] Run git ["fetch", "org", "contributor_patches"] in {REDACTED}
INFO [parity_processbot::cmd] Run git ["checkout", "-b", "contributor_patches_tmp"] in {REDACTED}
INFO [parity_processbot::cmd] Run git ["branch", "-D", "contributor_patches"] in {REDACTED}
INFO [parity_processbot::cmd] Run git ["checkout", "--track", "org/contributor_patches"] in {REDACTED}
INFO [parity_processbot::cmd] Run git ["branch", "-D", "contributor_patches_tmp"] in {REDACTED}
INFO [parity_processbot::cmd] Run git ["fetch", "origin", "master"] in {REDACTED}
INFO [parity_processbot::cmd] Run git ["rev-parse", "origin/master"] in {REDACTED}
INFO [parity_processbot::cmd] Run git ["merge", "origin/master", "--no-ff", "--no-edit"] in {REDACTED}
INFO [parity_processbot::cmd] Run git ["push", "org", "contributor_patches"] in {REDACTED}
INFO [parity_processbot::utils] Getting the head SHA after a repository update in org/contributor_patches
INFO [parity_processbot::cmd] Run git ["rev-parse", "HEAD"] in {REDACTED}
INFO [parity_processbot::webhook] https://github.com/org/owner/pull/1 merged successfully.

